android/pytorch_android/src/main/java/com/facebook/pytorch/Tensor.java

```diff
@@ -0,0 +1,320 @@
+    int pos = 0;
+    int numel = 1;
+    for (int i = dims.length  - 1; i >= 0; i--) {
+      pos += numel * index[i];
+      numel *= dims[i];
+    }
+    return pos;
+  }
+
+  public float[] getDataAsFloatArray() {
+    throw new IllegalStateException("Tensor of type " + getClass().getSimpleName() + " cannot " +
+        "return data as float.");
+  }
+
+  abstract int getTypeCode();
+
+  Buffer getRawDataBuffer() {
+    throw new IllegalStateException("Tensor of type " + getClass().getSimpleName() + " cannot " +
+        "return raw data buffer.");
+  }
+
+  private static String invalidIndexErrorMessage(int[] index, int dims[]) {
+    return String.format(Locale.US, "Invalid index %s for tensor dimensions %s",
+        Arrays.toString(index), Arrays.toString(dims));
+  }
+
+  static class Tensor_float32 extends Tensor {
+    private final FloatBuffer data;
+
+    Tensor_float32(FloatBuffer data, int[] dims) {
+      super(dims);
+      this.data = data;
+    }
+
+    @Override
+    public float[] getDataAsFloatArray() {
+      data.rewind();
+      float[] arr = new float[data.remaining()];
+      data.get(arr);
+      return arr;
+    }
+
+    @Override
+    int getTypeCode() {
+      return TYPE_CODE_FLOAT32;
+    }
+
+    @Override
+    Buffer getRawDataBuffer() {
+      return data;
+    }
+
+    @Override
+    public String toString() {
+      return String.format("Tensor_float32{dims:%s data:%s}", Arrays.toString(dims),
+          Arrays.toString(getDataAsFloatArray()));
+    }
+  }
+
+  static class Tensor_int32 extends Tensor {
+    private final IntBuffer data;
+
+    private Tensor_int32(IntBuffer data, int[] dims) {
+      super(dims);
+      this.data = data;
+    }
+
+    @Override
+    int getTypeCode() {
+      return TYPE_CODE_INT32;
+    }
+
+    @Override
+    Buffer getRawDataBuffer() {
+      return data;
+    }
+
+    public int[] getDataAsIntArray() {
+      data.rewind();
+      int[] arr = new int[data.remaining()];
+      data.get(arr);
+      return arr;
+    }
+
+    @Override
+    public String toString() {
+      return String.format("Tensor_int32{dims:%s data:%s}", Arrays.toString(dims),
+          Arrays.toString(getDataAsIntArray()));
+    }
+  }
+
+  static class Tensor_byte extends Tensor {
+    private final ByteBuffer data;
+
+    private Tensor_byte(ByteBuffer data, int[] dims) {
+      super(dims);
+      this.data = data;
+    }
+
+    @Override
+    int getTypeCode() {
+      return TYPE_CODE_BYTE;
+    }
+
+    @Override
+    Buffer getRawDataBuffer() {
+      return data;
+    }
+
+    public byte[] getDataAsByteArray() {
+      data.rewind();
+      byte[] arr = new byte[data.remaining()];
+      data.get(arr);
+      return arr;
+    }
+
+    @Override
+    public String toString() {
+      return String.format("Tensor_byte{dims:%s data:%s}", Arrays.toString(dims),
+          Arrays.toString(getDataAsByteArray()));
+    }
+  }
+
+  //region checks
```

Comment:
Does something consume this?

Reply:
Yes, idea allows to collapse //region...//endregion block